staubli_val3_driver/val3/ros_server/decTrajPtFull.pgx

```diff
@@ -32,54 +32,65 @@
 
   // robotId
   l_nRetVal=fromBinary(rosTrajPtFMsg.body.nData,4,"-4l",rosTrajPtFMsg.jointTrajPtFull.nRobotId)
-  if (l_nRetVal!=1)
+  if(l_nRetVal!=1)
     x_bFlag=false
     return
   endIf
 
   // sequence
   l_nRetVal=fromBinary(rosTrajPtFMsg.body.nData,4,"-4l",rosTrajPtFMsg.jointTrajPtFull.nSeq)
-  if (l_nRetVal!=1)
+  if(l_nRetVal!=1)
     x_bFlag=false
     return
   endIf
 
   // valid fields
   l_nRetVal=fromBinary(rosTrajPtFMsg.body.nData,4,"-4l",rosTrajPtFMsg.jointTrajPtFull.nValidFields)
-  if (l_nRetVal!=1)
+  if(l_nRetVal!=1)
     x_bFlag=false
     return
   endIf
 
   // time
   l_nRetVal=fromBinary(rosTrajPtFMsg.body.nData,4,"4.0l",rosTrajPtFMsg.jointTrajPtFull.nTime)
-  if (l_nRetVal!=1)
+  if(l_nRetVal!=1)
     x_bFlag=false
     return
   endIf
 
   // positions[10]
   for l_nIndex=0 to 9 step 1
     l_nRetVal=fromBinary(rosTrajPtFMsg.body.nData[(l_nIndex*4)+16],4,"4.0l",rosTrajPtFMsg.jointTrajPtFull.nPositions[l_nIndex])
-    if (l_nRetVal!=1)
+    if(l_nRetVal!=1)
       x_bFlag=false
       return
     endIf
   endFor
 
   // velocities[10]
   for l_nIndex=0 to 9 step 1
+    // Velocity buffer to smooth out last traj point
+    nVel[1] = nVel[0]
+    nVel[0] = rosTrajPtMsg.jointTrajPt.nVelocity
     l_nRetVal=fromBinary(rosTrajPtFMsg.body.nData[(l_nIndex*4)+56],4,"4.0l",rosTrajPtFMsg.jointTrajPtFull.nVelocities[l_nIndex])
-    if (l_nRetVal!=1)
+    if(l_nRetVal!=1)
       x_bFlag=false
       return
     endIf
+    // Round to remove trailing decimal places from fromBinary()
+    rosTrajPtMsg.jointTrajPt.nVelocity = rosTrajPtMsg.jointTrajPt.nVelocity * 10000
+    rosTrajPtMsg.jointTrajPt.nVelocity = round(rosTrajPtMsg.jointTrajPt.nVelocity)
+    rosTrajPtMsg.jointTrajPt.nVelocity = rosTrajPtMsg.jointTrajPt.nVelocity / 10000
+    // Smooth last point in traj
+    if((rosTrajPtMsg.jointTrajPt.nVelocity == 0.1) and (rosTrajPtMsg.jointTrajPt.nSeq != 0))
+      rosTrajPtMsg.jointTrajPt.nVelocity = (nVel[0] - (nVel[1] - nVel[0]))
+    endIf
```

Comment:
This seems duplicated from `decTrajPt`. Would it be possible to refactor this to avoid the duplication?

Reply:
Sure, it could be separated into its own program if that makes more sense

Reply:
See my higher level comment about perhaps doing this at a different level of abstraction (or at least not as part of the decoding sequence of a message).